### PR TITLE
Feature/initial implementation

### PR DESCRIPTION
**Pull request by Soohyeuk**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+[tool.coverage.xml]
+output = "coverage.xml"
+
+[tool.mypy]
+python_version = "3.8"
+warn_return_any = true
+warn_unused_configs = true
+disallow_untyped_defs = true
+check_untyped_defs = true
+exclude = ["test_.*\\.py$"]  # Exclude all test files
```
Comment (jaydaftari) [inline pyproject.toml:52]: why excluded all the test files from mypy? Give reason??

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+testpaths = [
+    "tests/integration",
+    "tests/end_to_end",
+    "src/calculator",
+    "src/logger",
+    "src/notifier"
+]
+python_files = ["test_*.py"]
+
+[tool.coverage.run]
```
Comment (jaydaftari) [inline pyproject.toml:32]: Why are they excluded from test coverage, leave one line comment.

File: src/calculator/pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+[tool.mypy]
+strict = true
+
+[build-system]
+requires = ["setuptools>=42.0.0"]
+build-backend = "setuptools.build_meta"
+
+[tool.setuptools.packages.find]
+where = ["."]
+exclude = ["tests*", "test_*.py"]
```
Comment (jaydaftari) [inline src/calculator/pyproject.toml:33]: Here as well, why excluding it? Reason?

File: README.md [hunk 4]
```diff
@@ -1,2 +1,47 @@
...
+   MacOS and Linux:
+   ```bash
+    curl -LsSf https://astral.sh/uv/install.sh | sh
+    ```
+   Windows:
+   ```bash
+    powershell -ExecutionPolicy ByPass -c "irm https://astral.sh/uv/install.ps1 | iex"
+   ```
+4. Install dependencies:
+    uv pip install pytest pytest-cov ruff mypy coverage
```
Comment (jaydaftari) [inline README.md:36]: ​Why not use uv add to install any additional dependencies needed for development? User can direclty run uv sync (similar to pip install -r requirements.txt)? Please refer to this discussion for more details:"https://github.com/astral-sh/uv/issues/9219"

File: pyproject.toml (hunk 1 above)
Comment (Soohyeuk) [reply pyproject.toml:52]: I was thinking only the code for the component itself (which excluded test files in my decision) should be considered for mypy.

Comment (jaydaftari) [reply pyproject.toml:52]: No, you shouldn't exclude it, test cases should also follow the rules of static type checker. This implies for both ruff and mypy

File: tests/integration/test_calc_logger_integration.py
Comment (kc34251) [inline tests/integration/test_calc_logger_integration.py:1]: During integration testing, the unused components should be mocked in the test. In this case, there should be a mock notifier to simulate the notifier that will eventually be added in e2e flow. It may not matter much in this template repository as the 3 components are not very complex, but I believe that it is good practice to mock components that will be used.

File: tests/integration/test_calc_notifier_integration.py
Comment (kc34251) [inline tests/integration/test_calc_notifier_integration.py:1]: Same as above comment on integration testing, there should be a mock calculator in this test so that we are making sure that component communication is working as intended for the components that we are testing.

Comment (kc34251) [inline tests/integration/test_calc_notifier_integration.py:1]: File name also needs to be changed, I believe that test_calc_notifier is misleading as this file is actually testing logger_notifier integration.

File: src/calculator/test_calculator.py
Comment (kc34251) [inline src/calculator/test_calculator.py:1]: I believe that division should also be tested in the calculator unit test as it is included as part of the calculator implementation.

File: component.md
Comment (kc34251) [inline component.md:1]: I believe that this component.md file could be split into readme files that could be placed into each of the component subdirectories. The components are being described in a top-level directory but I believe that it would be more intuitive for each component subdirectory to have a description of what the component is and what it does.

**Review by kc34251 — COMMENTED** [review]
I believe that you guys have completed a solid foundation in building this template repository. There is very good directory structure with each component having its own subdirectory within src. The unit tests are separated from the overall test folder and are closer to each component that they correspond to. However there are some changes that I believe could make this repository better and more intuitive to work with. I believe that interfaces and dependency injection should be used within your components so that they are more easily adjustable as well as more modular. This could be done by separating your implementation files and protocol files within each component subdirectory as can be seen in Professor Yotov's example directory. I also believe that you should change your ruff checks on your code or fix issues within your coding style as your ruff check is currently failing in your CircleCI pipeline with code formatting errors and warnings. However, you could disable these checks if you want to stick with your current code format.

File: src/calculator/calculator.py
Comment (Yolo1105) [inline src/calculator/calculator.py:1]: The divide function in calculator.py raises a ValueError when dividing by zero. It might be more comphrensive to include a custom error message or define a specific exception for better error handling and debugging

File: src/calculator/test_calculator.py
Comment (Yolo1105) [inline src/calculator/test_calculator.py:1]: While the addition, subtraction, and multiplication tests cover basic cases, It might be helpful if consider adding tests for edge cases like extremely large or small numbers and floating-point precision errors

File: src/logger/logger.py
Comment (Yolo1105) [inline src/logger/logger.py:1]: The functions within each component are well-structured, but consider adding a brief docstring at the beginning of each function explaining input types, output types, and a short description of the functionality

File: README.md
Comment (Yolo1105) [inline README.md:1]: I notice the project includes optional dependencies like types-setuptools, consider adding a brief section in the README explaining when and why to install these dependencies

File: .circleci/config.yml
Comment (Yolo1105) [inline .circleci/config.yml:1]: while the config.yml covers all the successful test runs, it would be beneficial to include a dedicated test stage for expected failures, such as input validation errors or edge cases like division by zero in the Calculator component that mention before

File: tests/integration/test_calc_logger_integration.py
Comment (Yolo1105) [inline tests/integration/test_calc_logger_integration.py:1]: In the integration tests, ensure that each test is isolated from others. Currently, the real instances of components like Logger are used directly, which may lead to shared state issues. Consider using mocks like MagicMock from unittest.mock to isolate component behaviors and clear shared resources like logs after each test to maintain test independence

File: README.md
Comment (Yolo1105) [inline README.md:1]: Please include links to both successful and failed CircleCI runs as required. Additionally, ensure that the coverage report link is accessible directly from the CircleCI output to facilitate quick verification

File: .circleci/config.yml
Comment (Yolo1105) [inline .circleci/config.yml:1]: Please include links to both successful and failed CircleCI runs as required. Additionally, ensure that the coverage report link is accessible directly from the CircleCI output to facilitate quick verification

**Comment by ivanearisty** [comment]
The professor mentioned a preference for dependency injection and a clear separation between interface and implementation. Currently, components like the Calculator only offer functions without any configurable dependency. You might enhance the design by creating interfaces for external dependencies (e.g., a logging interface injected into the calculator) so that each component can be reused or replaced easily. Albeit, he did not specifically require dependency injection, but it was highly recommended.

In tandem with the point above, your tests explicitly define the routes to components. If you would have used dependency injection, you could import a calculator protocol and let uv give you the calculator implementation. The current setup makes tests be less extensible, and, if your physical structure where to change, could lead to some issues.

Also, one of the guidelines is that running a build script should not execute any tests. Consider adding or documenting a separate build target that builds the project without running tests.

Edit: 
Look into uv workspaces to add dependency injection.
Look into uv run scripts to fix build scripts.

You can ignore, but these are some notes from the positives i see in the template :)
Separating the components into folders physically.
Each component has its own tests and a dedicated pyproject.toml.
The project includes unit tests placed near the component code.
uv usage to create isolated environments and install dependencies is great. We did not think about adding uv, pytest, coverage… etc… as dependencies for the individual components, but it makes a lot of sense to have them. Especially since components run unit tests and might have their own ruff config rules. Good job! 
Your ruff “ignores” makes sense.
The CI/CD configuration with CircleCI is robust, and includes all the tools you use (ruff, mypy, and tests), though I am unsure whether you intentionally are failing lining or not?
component.md 
Global pyproject.toml for consolidation of build, lint, type check, and tests is modern and clean.

File: tests/integration/test_calc_notifier_integration.py
Comment (Shounak-Ghosh) [reply tests/integration/test_calc_notifier_integration.py:1]: I don't see any notifier testing in this file, could you please elaborate?

File: src/calculator/calculator.py
Comment (Shounak-Ghosh) [reply src/calculator/calculator.py:1]: Why
